xfer/pipes.go

```diff
@@ -0,0 +1,143 @@
+package xfer
+
+import (
+	"io"
+	"log"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+)
+
+// Pipe is a bi-directional channel from someone thing in the probe
+// to the UI.
+type Pipe interface {
+	ID() string
+	Port() io.ReadWriter
+	Starboard() io.ReadWriter
+	CopyToWebsocket(io.ReadWriter, *websocket.Conn) error
+
+	Close(time.Time)
+	Closed() bool
+	ClosedAt() time.Time
+	OnClose(func())
+}
+
+type pipe struct {
+	mtx             sync.Mutex
+	id              string
+	port, starboard io.ReadWriter
+	tombstone       time.Time
+	quit            chan struct{}
+	onClose         func()
+}
+
+// NewPipe makes a new... pipe.
+func NewPipe(id string) Pipe {
+	r1, w1 := io.Pipe()
+	r2, w2 := io.Pipe()
+	return &pipe{
+		id: id,
+		port: struct {
+			io.Reader
+			io.Writer
+		}{
+			r1, w2,
+		},
+		starboard: struct {
+			io.Reader
+			io.Writer
+		}{
+			r2, w1,
+		},
+		quit: make(chan struct{}),
+	}
+}
+
+func (p *pipe) ID() string {
+	return p.id
+}
+
+func (p *pipe) Port() io.ReadWriter {
+	return p.port
+}
+
+func (p *pipe) Starboard() io.ReadWriter {
+	return p.starboard
+}
+
+func (p *pipe) Close(t time.Time) {
```

Comment:
If the ends are `ReadWriteCloser`s does that let us simplify any of the usage? It seems like it would *almost* let us use `io.Copy` for wiring it up to the websockets...